### DISCUSSION TITLE
Setting up smallstep with Yubikey

### DISCUSSION THREAD

**Discussion by nidkil** [smallstep/certificates#584] [General] [answered]
Hi,

I'm folllowing the blog post  [Build a Tiny Certificate Authority For Your Homelab](https://smallstep.com/blog/build-a-tiny-ca-with-raspberry-pi-yubikey/). In **Part 3: Configuring Your CA** I'm getting the following error when I try to start step-ca.

> authority.GetTLSCertificate: error creating certificate: command failed: smart card error 6a80: incorrect parameter in command data field

This is my configuration:

```
{
        "root": "/etc/step-ca/certs/root_ca.crt",
        "federatedRoots": [],
        "crt": "/etc/step-ca/certs/intermediate_ca.crt",
        "key": "yubikey:slot-id=9c",
        "kms": {
                "type": "yubikey",
                "pin": "123456"
        },
        "address": ":443",
```

Does anyone have an idea what the problem is?

Regards, nidkil

**Comment by tashian** [comment, marked as answer]
Hi nidkil, I haven't seen this error before.

Can you share the output of `ykman info`, which `step-ca` version you are running (and, did you build it yourself or install a binary?). Finally, which Linux distro and version are you running?

[Here's a related issue (unresolved) in another piv-go project](https://github.com/FiloSottile/yubikey-agent/issues/15)—this may be an upstream issue in go-piv/piv-go.

**Reply by nidkil** [reply]
@tashian thx for the quick respons. 

I have followed all the steps as specified in the blog post, only updating to the latest versions.

I'm running the following software:
- Ubuntu 21.04 Server arm64
- Go v1.16.4
- Yubikey Manager v4.0.0a1
- Smallstep CA v015.15
- Smallstep CLI v0.15.15

This is the output of `ykman info`:

```
Device type: YubiKey 5 NFC
Serial number: 14308509
Firmware version: 5.2.7
Form factor: Keychain (USB-A)
Enabled USB interfaces: OTP, FIDO, CCID
NFC transport is enabled.

Applications	USB    	NFC    
OTP     	Enabled	Enabled	
FIDO U2F	Enabled	Enabled	
OpenPGP 	Enabled	Enabled	
PIV     	Enabled	Enabled	
OATH    	Enabled	Enabled	
FIDO2   	Enabled	Enabled	
```
```

**Reply by tashian** [reply]
Thanks @nidkil. Can you also post the output of `ykman piv info`?

**Reply by nidkil** [reply]
@tashian here you are:

```
PIV version: 5.2.7
PIN tries remaining: 3
Management key algorithm: TDES
CHUID:	3019d4e739da739ced39ce739d836858210842108421c84210c3eb34103bb08b6c1ab6a678aa9b3099e1a20bd4350832303330303130313e00fe00
CCC: 	No data available.
Slot 9a:
	Algorithm:	ECCP256
	Subject DN:	CN=Nidkil CA Root CA,O=Nidkil CA
	Issuer DN:	CN=Nidkil CA Root CA,O=Nidkil CA
	Serial:		63955509087441653642227933518625758798
	Fingerprint:	3419aab119b1512276b4db03ae083569b7592668b7cf023ad15a47c1740c887d
	Not before:	2021-05-24 19:05:38
	Not after:	2031-05-22 19:05:38
Slot 9c:
	Algorithm:	ECCP256
	Subject DN:	CN=Nidkil CA Intermediate CA,O=Nidkil CA
	Issuer DN:	CN=Nidkil CA Root CA,O=Nidkil CA
	Serial:		166935707878256431136754202625511245906
	Fingerprint:	02534d645d6301c118126d0f6c57c498c6d6602da2df6eebb7b4d06d252a14d0
	Not before:	2021-05-24 19:05:38
	Not after:	2031-05-22 19:05:38
```

**Reply by tashian** [reply]
This all looks great. Could you try this alternative block for "kms" configuration? I'm using the default management key and pin here, but if you've changed those values then you'll need to change it here:

```json
    "kms": {
        "type": "yubikey",
        "uri": "yubikey:management-key=010203040506070801020304050607080102030405060708?pin-value=123456"
    }
```

**Reply by nidkil** [reply]
Unfortunately I'm getting a different error :-(

>  authority.GetTLSCertificate: error creating certificate: command failed: smart card error 6a80: incorrect parameter in command data field

**Reply by nidkil** [reply]
@tashian is there away to enable debugging to see what is happening?

**Reply by tashian** [reply]
Yes, you can run `step-ca` with `STEPDEBUG=1` to enable debugging output.
For PIV, I believe `step-ca` will call the [pcscd](https://linux.die.net/man/8/pcscd) daemon, and you can enable debugging on pcsdc with `-d`

**Reply by nidkil** [reply]
I've set debug with:

```
export STEPDEBUG=1
```

Not much debugging info showing... The same output is shown if I set the value to 1 or 0. I'm I doing it correctly?

**Reply by nidkil** [reply]
@tashian I wiped the system and ran through the whole setup again. Using the exact same versions as your blog post. And it worked.

I wiped the system again and ran through the setup using the latest versions. And it worked :+1: 

Also tried the alternative `kms` entry with `uri` in the `/etc/step-ca/config/ca.json` and it worked. 

No idea what I did wrong in the first setup.

Thx for your help!

**Comment by nidkil** [comment]
On a side note, I setup `yubikey-agent` and I'm getting the same error.

> sign_and_send_pubkey: signing failed for ECDSA "YubiKey #14308549 PIV Slot 9a" from agent: agent refused operation

BTW this is on a Manjaro machine.

**Reply by tashian** [reply]
Interesting. Well, I'm stumped on this one. It sounds like this is an upstream issue with either go-piv or pcsc-lite. I'm sorry you haven't been able to get it working. If you have a different YubiKey you can try it with, that may make a difference.

**Reply by nidkil** [reply]
Okay my bad, it is working. I needed to tap the Yubikey :-)

**Reply by CvW** [reply]
Glad you got it working. 

FWIW, I found it significantly  easier to use the Yubikey application to set the PKI and PUK, getting those saved in a password manager, etc. before setting up the rest of the internal CA. While operating the various yubikey-related commands from the CLI is a good thing to know, a GUI makes the job somewhat less error-prone for beginners like me.
